### PR TITLE
.NET 9

### PR DESCRIPTION
**Pull request by chenzhitong**
update to .net 9

### CODE REVIEW

File: src/Neo.VM/ScriptBuilder.cs
Comment (shargon) [inline src/Neo.VM/ScriptBuilder.cs:129]: Data Length == 0 is empty too?

File: src/Plugins/TokensTracker/Extensions.cs
Comment (Jim8y) [inline src/Plugins/TokensTracker/Extensions.cs:31]: is empty can be used on null as well?

File: tests/Neo.VM.Tests/UT_Utility.cs
Comment (Jim8y) [inline tests/Neo.VM.Tests/UT_Utility.cs:57]: why is this change? the behavor is no longer the same?

File: tests/Neo.Extensions.Tests/UT_ByteExtensions.cs
Comment (Jim8y) [inline tests/Neo.Extensions.Tests/UT_ByteExtensions.cs:28]: does this mean Net9.0 has changed the exception type?

Comment (chenzhitong) [reply tests/Neo.Extensions.Tests/UT_ByteExtensions.cs:28]: yes

File: tests/Neo.VM.Tests/UT_Utility.cs
Comment (chenzhitong) [reply tests/Neo.VM.Tests/UT_Utility.cs:57]: `BigInteger.One << 32 << int.MaxValue` will throw OverflowException in .net 9

File: tests/Neo.Extensions.Tests/UT_ByteExtensions.cs
Comment (nan01ab) [reply tests/Neo.Extensions.Tests/UT_ByteExtensions.cs:28]: > does this mean Net9.0 has changed the exception type?

`ToHexString` uses `Convert.ToHexStrongLower` when run with NET.9.
May need to add an argument null-check to keep same semantics.

File: tests/Neo.VM.Tests/UT_Utility.cs
Comment (shargon) [reply tests/Neo.VM.Tests/UT_Utility.cs:57]: Then this test is useless

**Comment by cschuchardt88** [comment]
I rather not go to `dotnet` `9.0` it removes a lot of stuff and changes the way things work. and community is not to happy about it. Also its a `Short Term Release`.

**Comment by chenzhitong** [comment]
> I rather not go to `dotnet` `9.0` it removes a lot of stuff and changes the way things work. and community is not to happy about it. Also its a `Short Term Release`.
> 
> Also you missed `global.json`

I didn't miss global.json

```c#
{
  "sdk": {
    "version": "9.0.100",
    "rollForward": "latestFeature",
    "allowPrerelease": false
  }
}
```

I’m not sure whether to upgrade to .NET 9. I'll leave it up to you core developers to decide!

**Comment by vncoelho** [comment]
I am in favor, next year we move to long term again and so on.

**Comment by shargon** [comment]
> I am in favor, next year we move to long term again and so on.

me too

**Comment by cschuchardt88** [comment]
This will break the functionally of `RestServer`

**Comment by cschuchardt88** [comment]
> I’m not sure whether to upgrade to .NET 9. I'll leave it up to you core developers to decide!

No dont leave it up to them. If it's dotnet 9.0 then have to upgrade. there is no way around it.

**Review by cschuchardt88 — APPROVED** [review]

**Review by cschuchardt88 — COMMENTED** [review]
Why are you changing `*.cs` files?

**Comment by chenzhitong** [comment]
> Why are you changing `*.cs` files?

Because some of the exception types change, and some of the dotnet format rules change.

**Comment by shargon** [comment]
> > I’m not sure whether to upgrade to .NET 9. I'll leave it up to you core developers to decide!
> 
> No dont leave it up to them. If it's dotnet 9.0 then have to upgrade. there is no way around it.

That's true

File: src/Neo.VM/ScriptBuilder.cs
Comment (shargon) [inline src/Neo.VM/ScriptBuilder.cs:129]: What is this?

File: src/Plugins/TokensTracker/Extensions.cs
Comment (shargon) [inline src/Plugins/TokensTracker/Extensions.cs:32]: ?

File: src/Neo.VM/ScriptBuilder.cs
Comment (chenzhitong) [reply src/Neo.VM/ScriptBuilder.cs:129]: Because `if (data == null)` will fail the format check in .net 9.

File: src/Plugins/TokensTracker/Extensions.cs
Comment (chenzhitong) [reply src/Plugins/TokensTracker/Extensions.cs:32]: Because `item == null` will fail the format check in .net 9.

Comment (shargon) [reply src/Plugins/TokensTracker/Extensions.cs:32]: We should add test to ensure if IsEmpty check null and not Length != 0

Comment (chenzhitong) [reply src/Plugins/TokensTracker/Extensions.cs:32]: Running dotnet format automatically changes `item == null` to `item.IsEmpty`. If you use `item.IsEmpty`, the unit test will fail, I tried it. So I've changed it to `ReadOnlySpan<byte> nullData = null; item == nullData;`

Comment (cschuchardt88) [reply src/Plugins/TokensTracker/Extensions.cs:32]: just edit the `.editorconfig` and turn that off.

File: .editorconfig
Comment (cschuchardt88) [inline .editorconfig:29]: this should be on `line 58`

File: tests/Neo.ConsoleService.Tests/Neo.ConsoleService.Tests.csproj
Comment (cschuchardt88) [inline tests/Neo.ConsoleService.Tests/Neo.ConsoleService.Tests.csproj:19]: Please use the `Directory.Build.props` in the directory to update this.

**Review by cschuchardt88 — APPROVED** [review]

**Comment by shargon** [comment]
I would like to reduce the changes before merge it, and test this changes with the proper ut

File: src/Neo.CLI/CLI/MainService.cs
Comment (Jim8y) [inline src/Neo.CLI/CLI/MainService.cs:530]: this is the only place where code is changed in the core. @shargon

Comment (chenzhitong) [reply src/Neo.CLI/CLI/MainService.cs:530]: This paragraph was automatically changed after running `dotnet format` in dotnet 9, and I saw that the logic hadn't changed, so I submitted it.

**Review by Jim8y — DISMISSED** [review]

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs
Comment (Jim8y) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: the warning shows `Comparing a span to 'null' might be redundant, the 'null' literal will be implicitly converted to a 'Span<T>.Empty'`

File: src/Neo.CLI/CLI/MainService.cs
Comment (cschuchardt88) [reply src/Neo.CLI/CLI/MainService.cs:530]: turn it off in the `.editorconfig`

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs
Comment (shargon) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: If it worked like this in net8, we should remove the null check and push always empty array

Comment (shargon) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: But previously seems that with null we receive an error and with empty an empty push, not we can't

Comment (Jim8y) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: Tested new UT on master, it does not work as well

```
Test method Neo.Test.UT_ScriptBuilder.TestNullAndEmpty threw exception: 
System.ArgumentNullException: Value cannot be null. (Parameter 'data')
    at Neo.VM.ScriptBuilder.EmitPush(ReadOnlySpan`1 data) in C:\Users\liaoj\git\neo\src\Neo.VM\ScriptBuilder.cs:line 130
   at Neo.Test.UT_ScriptBuilder.TestNullAndEmpty() in C:\Users\liaoj\git\neo\tests\Neo.VM.Tests\UT_ScriptBuilder.cs:line 57
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)

```

Comment (shargon) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: Then we should push always empty array and avoid the exception

File: .editorconfig
Comment (chenzhitong) [inline .editorconfig]: Why did you change this back?

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs
Comment (Jim8y) [inline tests/Neo.VM.Tests/UT_ScriptBuilder.cs:52]:
```suggestion
                    ReadOnlySpan<byte> span = [];
```

**Review by Jim8y — DISMISSED** [review]

**Review by shargon — APPROVED** [review]

**Review by shargon — APPROVED** [review]

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs
Comment (cschuchardt88) [inline tests/Neo.VM.Tests/UT_ScriptBuilder.cs:50]: should be using `empty`, `null` doesn't really work for `spans`

File: src/Neo.VM/Types/Map.cs
Comment (cschuchardt88) [inline src/Neo.VM/Types/Map.cs:31]: Where is `Collections` pointing to; `system` or `vm`?

**Review by cschuchardt88 — APPROVED** [review]

File: src/Neo.VM/Types/Map.cs
Comment (Jim8y) [reply src/Neo.VM/Types/Map.cs:31]: VM

**Comment by shargon** [comment]
We need to revert it https://github.com/neo-project/neo-devpack-dotnet/pull/1257#issuecomment-2507022502

**Pull request merged**
